File: backend/src/routes/tenants.js
'use strict';
const express = require('express');
const { body, param } = require('express-validator');
const db = require('../config/db');
const { authenticate } = require('../middleware/auth');
const { roleGuard } = require('../middleware/roleGuard');
const { validate } = require('../middleware/validate');
const { writeAudit } = require('../services/auditService');

const router = express.Router();
const IT = ['IT_ADMIN', 'SUPERADMIN'];

router.get('/', authenticate, async (req, res) => {
  const { rows } = await db.query('SELECT * FROM tenants ORDER BY nazwa');
  res.json(rows);
});

router.get('/:id', authenticate, async (req, res) => {
  const { rows } = await db.query('SELECT * FROM tenants WHERE id = $1', [req.params.id]);
  if (!rows.length) return res.status(404).json({ error: 'Nie znaleziono jednostki.' });
  res.json(rows[0]);
});

router.post('/', authenticate, roleGuard(...IT),
  [
    body('nazwa').notEmpty().trim(),
    body('skrot').notEmpty().trim().isLength({ max: 20 }),
    body('dni_do_przegladu').optional().isInt({ min: 30 }),
  ],
  validate,
  async (req, res) => {
    const { nazwa, skrot, regon, nip, dni_do_przegladu = 365 } = req.body;
    const { rows } = await db.query(
      `INSERT INTO tenants (nazwa, skrot, regon, nip, dni_do_przegladu)
       VALUES ($1,$2,$3,$4,$5) RETURNING *`,
      [nazwa, skrot, regon || null, nip || null, dni_do_przegladu]
    );
    await writeAudit({ userId: req.user.userId, username: req.user.username, rola: req.user.rola, akcja: 'TENANT_CREATE', tabelaDocelowa: 'tenants', rekordId: rows[0].id, noweDane: rows[0] });
    res.status(201).json(rows[0]);
  }
);

router.put('/:id', authenticate, roleGuard(...IT),
  [body('nazwa').notEmpty().trim()],
  validate,
  async (req, res) => {
    const { nazwa, skrot, regon, nip, aktywny, dni_do_przegladu } = req.body;
    const { rows } = await db.query(
      `UPDATE tenants SET nazwa=$1, skrot=$2, regon=$3, nip=$4, aktywny=$5, dni_do_przegladu=$6
       WHERE id=$7 RETURNING *`,
      [nazwa, skrot, regon || null, nip || null, aktywny !== undefined ? aktywny : true, dni_do_przegladu || 365, req.params.id]
    );
    if (!rows.length) return res.status(404).json({ error: 'Nie znaleziono jednostki.' });
    await writeAudit({ userId: req.user.userId, username: req.user.username, rola: req.user.rola, akcja: 'TENANT_UPDATE', tabelaDocelowa: 'tenants', rekordId: rows[0].id, noweDane: rows[0] });
    res.json(rows[0]);
  }
);

router.delete('/:id', authenticate, roleGuard('SUPERADMIN'), async (req, res) => {
  await db.query('UPDATE tenants SET aktywny=FALSE WHERE id=$1', [req.params.id]);
  await writeAudit({ userId: req.user.userId, username: req.user.username, rola: req.user.rola, akcja: 'TENANT_DEACTIVATE', tabelaDocelowa: 'tenants', rekordId: req.params.id });
  res.json({ message: 'Jednostka dezaktywowana.' });
});

module.exports = router;

File: backend-py/routes/tenants.py
from __future__ import annotations
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy import text
from sqlalchemy.orm import Session

from database import get_session
from dependencies import CurrentUser, get_current_user, require_roles, write_audit, get_client_ip
from fastapi import Request

router = APIRouter()
IT = ("IT_ADMIN", "SUPERADMIN")


class TenantIn(BaseModel):
    nazwa: str
    skrot: str
    regon: Optional[str] = None
    nip: Optional[str] = None
    aktywny: Optional[bool] = True
    dni_do_przegladu: Optional[int] = 365


@router.get("")
async def list_tenants(session: Session = Depends(get_session),
                       current_user: CurrentUser = Depends(get_current_user)):
    rows = session.execute(text("SELECT * FROM tenants ORDER BY nazwa")).mappings().all()
    return [dict(r) for r in rows]


@router.get("/{tenant_id}")
async def get_tenant(tenant_id: int, session: Session = Depends(get_session),
                     current_user: CurrentUser = Depends(get_current_user)):
    row = session.execute(text("SELECT * FROM tenants WHERE id=:id"), {"id": tenant_id}).mappings().first()
    if not row:
        raise HTTPException(status_code=404, detail="Nie znaleziono jednostki.")
    return dict(row)


@router.post("", status_code=201)
async def create_tenant(body: TenantIn, request: Request,
                        session: Session = Depends(get_session),
                        current_user: CurrentUser = Depends(require_roles(*IT))):
    from datetime import datetime
    row = session.execute(
        text("INSERT INTO tenants (nazwa, skrot, regon, nip, dni_do_przegladu, aktywny, data_utworzenia) VALUES (:n,:s,:r,:nip,:d,1,:dt) RETURNING *"),
        {"n": body.nazwa, "s": body.skrot, "r": body.regon, "nip": body.nip,
         "d": body.dni_do_przegladu or 365, "dt": datetime.utcnow()},
    ).mappings().first()
    session.commit()
    result = dict(row)
    write_audit(session, user_id=current_user.userId, username=current_user.username,
                rola=current_user.rola, akcja="TENANT_CREATE", tabela_docelowa="tenants",
                rekord_id=result["id"], nowe_dane=result, ip_adres=get_client_ip(request))
    return result


@router.put("/{tenant_id}")
async def update_tenant(tenant_id: int, body: TenantIn, request: Request,
                        session: Session = Depends(get_session),
                        current_user: CurrentUser = Depends(require_roles(*IT))):
    row = session.execute(
        text("UPDATE tenants SET nazwa=:n, skrot=:s, regon=:r, nip=:nip, aktywny=:a, dni_do_przegladu=:d WHERE id=:id RETURNING *"),
        {"n": body.nazwa, "s": body.skrot, "r": body.regon, "nip": body.nip,
         "a": body.aktywny if body.aktywny is not None else True,
         "d": body.dni_do_przegladu or 365, "id": tenant_id},
    ).mappings().first()
    session.commit()
    if not row:
        raise HTTPException(status_code=404, detail="Nie znaleziono jednostki.")
    result = dict(row)
    write_audit(session, user_id=current_user.userId, username=current_user.username,
                rola=current_user.rola, akcja="TENANT_UPDATE", tabela_docelowa="tenants",
                rekord_id=tenant_id, nowe_dane=result, ip_adres=get_client_ip(request))
    return result


@router.delete("/{tenant_id}")
async def delete_tenant(tenant_id: int, request: Request,
                        session: Session = Depends(get_session),
                        current_user: CurrentUser = Depends(require_roles("SUPERADMIN"))):
    session.execute(text("UPDATE tenants SET aktywny=0 WHERE id=:id"), {"id": tenant_id})
    session.commit()
    write_audit(session, user_id=current_user.userId, username=current_user.username,
                rola=current_user.rola, akcja="TENANT_DEACTIVATE", tabela_docelowa="tenants",
                rekord_id=tenant_id, ip_adres=get_client_ip(request))
    return {"message": "Jednostka dezaktywowana."}
